### PR TITLE
Changed setValue for the special case of value is None

### PR DESCRIPTION
**Pull request by jakirkham**
If value is None, then this should have the effect of making the slot unready. In addition, notifications should be sent if the state changed. These will be either ready or unready notifications based on the value given.

### CODE REVIEW

**Comment by jakirkham** [comment]
I should add that using setValue with None previously had undefined behavior. In other words, it raised an exception and left the slot in an undefined state. This fixes that problem.

**Comment by jakirkham** [comment]
Rebased on current master.

**Comment by jakirkham** [comment]
Added tests to verify that setting an `InputSlot` to `None` works.

Note that these tests would have failed previously due to indeterminate behavior. In other words, setting the value would work; however, the slot would still claim to be ready and would still try to retrieve a value and fail with the message below.

    AssertionError: This inputSlot has no value and no partner.  You can't ask for its data yet!`

In other words, this pull request not only provides a feature, but fixes a serious bug.

**Comment by jakirkham** [comment]
@stuarteberg, can you please look at this again when you have a chance?
